### PR TITLE
Add more JSON

### PR DESCRIPTION
**Pull request by marler8997**
Added fields to `compilerInfo` object in JSON object.

Note that this is based on https://github.com/dlang/dmd/pull/7521 

It differs a bit, electing to go for a json representation matching the internals rather than trying to exactly match DUB probe files.  Not sure which approach is better for DMD, this one is easier to implement so I went with this one for now.

Example Output:
```json
{
 "compilerInfo" : {
  "vendor" : "Digital Mars D",
  "version" : "v2.078.2-742-gd6b8376-dirty",
  "__VERSION__" : 2078,
  "compilerInterface" : "dmd",
  "bitWidth" : 64,
  "platforms" : [
   "posix"
  ],
  "architectures" : [
   "x86_64"
  ],
  "predefinedVersions" : [],
  "supportsIncludeImports" : true
 },
 "buildInfo" : {
  "cwd" : "/home/marlerj/ddev1/dmd",
  "argv0" : "./generated/linux/release/64/dmd",
  "config" : "./generated/linux/release/64/dmd.conf",
  "importPaths" : [
   "./generated/linux/release/64/../../../../../druntime/import",
   "./generated/linux/release/64/../../../../../phobos"
  ],
  "objectFiles" : [],
  "libraryFiles" : [],
  "ddocFiles" : [],
  "mapFile" : null,
  "resourceFile" : null,
  "defFile" : null
 }
}
```

### CODE REVIEW

**Review by wilzbach — COMMENTED** [review]
Wow that was quick! Thanks!
~~~`config` and `compilerFrontend` are globals, so they could be easily added from this PR.~~~

`compiler` and `frontendVersion` are a bit more effort, so it probably makes sense to do them in a quick followup (I can do this too).

Otherwise LGTM!

**Comment by wilzbach** [comment]
Ah I see:
- `config` is already part of `buildInfo`
-  `version` is part of `compilerInfo`

File: src/dmd/json.d [hunk 1]
```diff
@@ -826,8 +826,60 @@ public:
     private void generateCompilerInfo()
     {
         objectStart();
-        property("binary", global.params.argv0);
+        property("vendor", global.compiler.vendor);
         property("version", global._version);
+        property("argv0", global.params.argv0);
```
[the remark this answers is not in the document]
Comment (marler8997) [reply src/dmd/json.d]: Well, I was thinking that `argv0` is more true to what it actually is.  "binary" sounds more like a filename but argv0 can be an alias/string/filename.  It may be helpful to include both "binary" and "argv0"...

File: src/dmd/json.d [hunk 2]
```diff
@@ -826,8 +826,60 @@ public:
     private void generateCompilerInfo()
     {
         objectStart();
-        property("binary", global.params.argv0);
+        property("vendor", global.compiler.vendor);
         property("version", global._version);
```
[the remark this answers is not in the document]
Comment (marler8997) [reply src/dmd/json.d]: inifilename is already in the `buildInfo` field

If we add DFLAGS...wouldn't we want to add all the command line arguments?

File: src/dmd/json.d [hunk 3]
```diff
@@ -826,8 +826,60 @@ public:
...
+        property("argv0", global.params.argv0);
+
```
[the remark this answers is not in the document]
Comment (marler8997) [reply src/dmd/json.d]: Can you elaborate what `frontendVersion` and `compilerFrontend` are for?

File: src/dmd/json.d (hunk 2 above)
[the remark this answers is not in the document]
Comment (wilzbach) [reply src/dmd/json.d]: > If we add DFLAGS...wouldn't we want to add all the command line arguments?

Yes, but again doesn't have to be in this PR.

File: src/dmd/json.d (hunk 3 above)
[the remark this answers is not in the document]
Comment (wilzbach) [reply src/dmd/json.d]: (these variables are currently used by DUB in its platform probing)

- `frontendVersion`: 2071, 2072, known in D as `__VERSION__` (and in DMD as `VERSIONX` token). This is important for feature and compatibility detection
- `compiler`, e.g. `dmd`, `ldc` or `gdc` - as mentioned the vendor might be different to the CLI interface (maybe `compilerInterface` is a better name)
- `compilerFrontend` - same as `version` (already existent)

**Comment by timotheecour** [comment]
definitely not PR blockers (could be done in future ones) but for reference:
* json output shd probably be somewhere in docs (maybe in https://dlang.org/dmd-osx.html) ; real useful to have
* linker flags
* example with architectures.length > 1 
* predefinedVersions shouldn’t be empty (bug?)
* stringImportPaths (on that note IIRC there was a bug where these would become absolute paths)

**Comment by marler8997** [comment]
Ok, I think this is a good start.  I've included "frontendVersion" as "__VERSION__" and also included "compilerInterface".  This should definitely have everything `rdmd` needs.  I didn't add `DFLAGS`...I think I'll let @wilzbach figure out the best way to add that.  Since this is experimental, if there's something wrong we can always fix it later.

Note for future changes.  I opted for the compilerInfo/buildInfo objects to ALWAYS include fields even if their values are null/empty.  This way a user can still see the format of the document even if a value is null (see `resourceFile`/`defFile`/etc as examples).  However, to have these semantics with string fields, you need to use the `propertyOrNull` function instead of `property`.  The `property` function will just leave the field out if the value is null.

**Comment by marler8997** [comment]
NOTE:  If you copy the following script to the test directory, you can run it after building DMD to automatically update the json tests with the new expected JSON output:

https://gist.github.com/marler8997/f3973cc6bcde7a88d04f48140ea8fe6b
```BASH
## Script to update json test files
set -x

make test_results/compilable/json2.d.out
cp test_results/compilable/json2.out.sanitized compilable/extra-files/json2.out

make test_results/compilable/jsonCompilerInfo.d.out
cp test_results/compilable/jsonCompilerInfo.out.sanitized compilable/extra-files/jsonCompilerInfo.out

make test_results/compilable/json_nosource.sh.out
cp test_results/compilable/json_nosource.out.sanitized compilable/extra-files/json_nosource.out
```

File: src/dmd/json.d [hunk 4]
```diff
@@ -826,8 +836,61 @@ public:
     private void generateCompilerInfo()
     {
         objectStart();
-        property("binary", global.params.argv0);
-        property("version", global._version);
+        propertyOrNull("vendor", global.compiler.vendor);
+        propertyOrNull("version", global._version);
+        property("__VERSION__", global.versionNumber());
+        propertyOrNull("interface", determineCompilerInterface());
+        property("bitWidth", size_t.sizeof * 8);
```
Comment (wilzbach) [inline src/dmd/json.d]: Not sure if `bitWidth` is a good name.

File: src/dmd/json.d [hunk 5]
```diff
@@ -867,9 +962,8 @@ public:
         foreach (m; Module.amodules)
         {
             objectStart();
-            if(m.md)
-                property("name", m.md.toChars());
-            property("file", m.srcfile.toChars());
+            propertyOrNull("name", m.md ? m.md.toChars() : null);
+            propertyOrNull("file", m.srcfile.toChars());
```
Comment (wilzbach) [inline src/dmd/json.d]: Can this even be `null`?

File: src/dmd/json.d [hunk 6]
```diff
@@ -968,3 +1062,20 @@ JsonFieldFlags tryParseJsonField(const(char)* fieldName)
...
     return JsonFieldFlags.none;
 }
+
+/**
+Determines and returns the compiler interface which is one of `dmd`, `ldc`,
+`gdc` or `sdc`. Returns `null` if no interface can be determined.
+*/
+private const(char)* determineCompilerInterface()
+{
+    if (0 == strcmp(global.compiler.vendor, "Digital Mars D"))
```
Comment (wilzbach) [inline src/dmd/json.d]: Pattern is `actual == expected`  (grep through the source code).
Of course `!strcmp` would work too.

File: src/dmd/globals.d [hunk 7]
```diff
@@ -477,3 +515,4 @@ enum PINLINE : int
 alias StorageClass = uinteger_t;
 
 extern (C++) __gshared Global global;
+
```
Comment (wilzbach) [inline src/dmd/globals.d]: Unrelated?

File: src/dmd/json.d [hunk 8]
```diff
@@ -839,18 +902,50 @@ public:
...
         objectStart();
-        property("cwd", getcwd(null, 0));
-        property("config", global.inifilename ? global.inifilename : null);
-        if (global.params.lib) {
-            property("library", global.params.libname);
-        }
+        propertyOrNull("cwd", getcwd(null, 0));
+        propertyOrNull("argv0", global.params.argv0);
+        propertyOrNull("config", global.inifilename);
+        propertyOrNull("libname", global.params.libname);
```
Comment (wilzbach) [inline src/dmd/json.d]: Everything else uses camelCase, might be a good idea to use `libName` too?
Precedent is above: we don't use `inifilename`, but `config`

File: src/dmd/json.d (hunk 4 above)
Comment (marler8997) [reply src/dmd/json.d]: yeah, any other ideas?  It's the same number that's printed in the --version output.

File: src/dmd/json.d (hunk 5 above)
Comment (marler8997) [reply src/dmd/json.d]: no, but it seemed like consistently using `propertyOrNull` was better.  The new objects should use `propertyOrNull` and the old `modules` object uses property....maybe there is a better name....

File: src/dmd/globals.d [hunk 9]
```diff
@@ -362,6 +362,44 @@ struct Global
...
         compiler.vendor = "Digital Mars D";
     }
+
+    /**
+    Returns: the version as the number that would be returned for __VERSION__
+    */
+    extern(C++) uint versionNumber()
+    {
+        import core.stdc.ctype;
+        static uint cached = 0;
```
Comment (WalterBright) [inline src/dmd/globals.d]: Don't use thread local storage for dmd. Use `__gshared`

File: src/dmd/json.d [hunk 10]
```diff
@@ -237,6 +237,16 @@ public:
         comma();
     }
 
+    void propertyOrNull(const(char)* name, const(char)* s)
```
Comment (WalterBright) [inline src/dmd/json.d]: No Ddoc header

File: src/dmd/json.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/json.d]: `size_tSizeof`

File: src/dmd/json.d [hunk 11]
```diff
@@ -826,8 +836,61 @@ public:
...
+        if (global.params.isWindows)
+        {
+            item("windows");
+        }
+        else
+        {
+            item("posix");
+            if (global.params.isLinux)
+                item("linux");
+            if (global.params.isOSX)
```
Comment (WalterBright) [inline src/dmd/json.d]: `else if`, etc.

File: src/dmd/globals.d (hunk 9 above)
Comment (marler8997) [reply src/dmd/globals.d]: good catch

File: src/dmd/json.d (hunk 4 above)
Comment (marler8997) [reply src/dmd/json.d]: ok, `size_t` seems reasonable.  With `size_t` I also changed it to the "byte-width" instead of "bit-width".   Let me know if someone prefers it to be the bit-width instead.

File: src/dmd/json.d (hunk 5 above)
Comment (marler8997) [reply src/dmd/json.d]: The name `requiredProperty` came to mind and seems like a good idea.

File: src/dmd/globals.d [hunk 12]
```diff
@@ -362,6 +362,44 @@ struct Global
         _version = (import("VERSION") ~ '\0').ptr;
         compiler.vendor = "Digital Mars D";
     }
+
+    /**
+    Returns: the version as the number that would be returned for __VERSION__
+    */
+    extern(C++) uint versionNumber()
```
Comment (WalterBright) [inline src/dmd/globals.d]: This code duplicates code in the lexer https://github.com/dlang/dmd/blob/master/src/dmd/lexer.d#L518

In fact, it's a cut&paste of it. It should take an argument instead of relying on a global, and moved to another module (lexer.d?).

Comment (marler8997) [reply src/dmd/globals.d]: As you can see later on in this PR, the code was removed from lexer.d and replaced with a call to this function.  I could move this function to lexer.d and have globals import lexer (instead of the other way around).  Would you prefer that?

Comment (marler8997) [reply src/dmd/globals.d]: Yes it could take an argument, but then the value won't be cached.  This current implementation allows the value to be cached in case `__VERSION__` is used multiple times along with outputing the `__VERSION__` json value.

Comment (WalterBright) [reply src/dmd/globals.d]: I see you are correct. I missed that. I with draw both suggestions.

**Comment by WalterBright** [comment]
Is this no longer a WIP?

**Comment by marler8997** [comment]
Correct no longer WIP.  Though this does not "finalize" all the JSON fields...this should be ready for merge so others can build upon it.
